src/Expander.hs

```diff
@@ -952,13 +1081,101 @@ runTask (tid, localData, task) = withLocal localData $ do
               \case
                 Nothing -> Just $ Env.singleton x n val
                 Just env -> Just $ env <> Env.singleton x n val
+    TypeCheck eDest ty -> do
+      st <- getState
+      let eTop = view (expanderCompletedCore . at eDest) st
+      case eTop of
+        Nothing -> stillStuck tid task
+        Just e ->
+          case ty of
+            IncompleteType tDest ->
+              linkedType tDest >>=
+              \case
+                Nothing -> stillStuck tid task
+                Just t ->
+                  typeCheckLayer eDest e t
+            CompleteType t ->
+              typeCheckLayer eDest e t
+    TypeCheckDecl dest -> do
+      st <- getState
+      case view (expanderCompletedDecls . at dest) st of
+        Nothing -> stillStuck tid task
+        Just d -> typeCheckDecl d
+    GeneralizeType edest ty schdest -> do
+      ready <- isExprChecked edest
+      if ready
+        then do
+          st <- getState
+          case view (expanderExpressionTypes . at edest) st of
+            Nothing -> throwError $ InternalError "Type not found during generalization"
+            Just _ -> do
+              sch <- generalizeType ty
+              linkScheme schdest sch
+        else stillStuck tid task
+    TypeCheckVar eDest ty ->
+      linkedCore eDest >>=
+      \case
+        Nothing -> stillStuck tid task
+        Just (Core (CoreVar x)) ->
+          varType x >>=
+          \case
+            Nothing -> stillStuck tid task
+            Just sch -> do
+              specialize sch >>= unify eDest ty
+              saveExprType eDest ty
+        Just _ ->
+          throwError $ InternalError "Not a variable when specializing"
+
   where
     laterMacro tid' b v x dest deps mdest stx = do
       localConfig <- view expanderLocal
       modifyState $
         over expanderTasks $
           ((tid', localConfig, AwaitingMacro dest (TaskAwaitMacro b v x deps mdest stx)) :)
 
+expandOneType :: SplitTypePtr -> Syntax -> Expand ()
+expandOneType dest stx
+  | Just ident <- identifierHeaded stx = do
+      b <- resolve =<< addRootScope ident
+      v <- getEValue b
+      case v of
+        EPrimTypeMacro impl -> impl dest stx
+        EPrimMacro _ -> throwError $ InternalError "Context expects a type"
```

Comment:
Could these errors arise from user macros in macro positions returning non-type expressions?

Reply:
Exactly.

Reply:
Then they aren't "internal errors" but rather "user errors" :-)

Reply:
I agree, and that's why it's now fixed :-)